### PR TITLE
Hotfix equals avoids null visitor

### PR DESCRIPTION
**Pull request by punkratz312**

### CODE REVIEW

File: src/test/java/org/openrewrite/staticanalysis/EqualsAvoidsNullTest.java [hunk 1]
```diff
@@ -45,7 +45,13 @@ public class A {
                       System.out.println(s.compareTo("test"));
                       System.out.println(s.compareToIgnoreCase("test"));
                       System.out.println(s.contentEquals("test"));
+                      System.out.println(s.contentEquals("test"));
```
Comment (punkratz312) [inline src/test/java/org/openrewrite/staticanalysis/EqualsAvoidsNullTest.java:48]:
```suggestion
```

**Comment by timtebeek** [comment]
Adding the failure from the logs
```
EqualsAvoidsNullTest > invertConditional() FAILED
    java.lang.IllegalStateException: LST contains missing or invalid type information
    Identifier->MethodInvocation->MethodInvocation->MethodInvocation->MethodInvocation->Block->Block->ClassDeclaration->CompilationUnit
    /*~~(Identifier type is missing or malformed)~~>*/Stream

    Identifier->MethodInvocation->Lambda->MethodInvocation->MethodInvocation->MethodInvocation->Block->Block->ClassDeclaration->CompilationUnit
    /*~~(Identifier type is missing or malformed)~~>*/item
    https://docs.openrewrite.org/reference/faq#im-seeing-lst-contains-missing-or-invalid-type-information-in-my-recipe-unit-tests-how-to-resolve
        at org.openrewrite.java.Assertions.assertValidTypes(Assertions.java:92)
        at org.openrewrite.java.Assertions.validateTypes(Assertions.java:60)
        at org.openrewrite.java.Assertions$$Lambda$815/0x00007fbc8047aee0.accept(Unknown Source)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:303)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:132)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:127)
        at org.openrewrite.staticanalysis.EqualsAvoidsNullTest.invertConditional(EqualsAvoidsNullTest.java:36)
```

**Comment by timtebeek** [comment]
Right now this fails on those commented out lines only it seems
```
EqualsAvoidsNullTest > invertConditional() FAILED
    org.opentest4j.AssertionFailedError: Unexpected result in "A.java":
```
```diff
diff --git a/A.java b/A.java
index 03e98ce..3043936 100644
--- a/A.java
+++ b/A.java
@@ -10,7 +10,7 @@ 
                       System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
                   }
 //                  boolean isFoo(final String test) {
-//                     return "test".contentEquals(test)
-//                             || "test2".contentEquals(test);
+//                     return test.contentEquals("test")
+//                             || test.contentEquals("test2");
 //                  }
               }
\ No newline at end of file
```
```
    expected: 
      "              import java.util.stream.Stream;
                    public class A {
                        {
                            String s = null;
                            if("test".equals(s)) {}
                            if("test".equalsIgnoreCase(s)) {}
                            System.out.println("test".compareTo(s));
                            System.out.println("test".compareToIgnoreCase(s));
                            System.out.println("test".contentEquals(s));
                            System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
                        }
      //                  boolean isFoo(final String test) {
      //                     return "test".contentEquals(test)
      //                             || "test2".contentEquals(test);
      //                  }
                    }"
     but was: 
      "              import java.util.stream.Stream;
                    public class A {
                        {
                            String s = null;
                            if("test".equals(s)) {}
                            if("test".equalsIgnoreCase(s)) {}
                            System.out.println("test".compareTo(s));
                            System.out.println("test".compareToIgnoreCase(s));
                            System.out.println("test".contentEquals(s));
                            System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
                        }
      //                  boolean isFoo(final String test) {
      //                     return test.contentEquals("test")
      //                             || test.contentEquals("test2");
      //                  }
                    }"
        at app//org.openrewrite.test.RewriteTest.assertContentEquals(RewriteTest.java:620)
        at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:509)
        at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:132)
        at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:127)
        at app//org.openrewrite.staticanalysis.EqualsAvoidsNullTest.invertConditional(EqualsAvoidsNullTest.java:36)

```

**Comment by punkratz312** [comment]
> Right now this fails on those commented out lines only it seems
> 
> ```
> EqualsAvoidsNullTest > invertConditional() FAILED
>     org.opentest4j.AssertionFailedError: Unexpected result in "A.java":
> ```
> 
> ```diff
> diff --git a/A.java b/A.java
> index 03e98ce..3043936 100644
> --- a/A.java
> +++ b/A.java
> @@ -10,7 +10,7 @@ 
>                        System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
>                    }
>  //                  boolean isFoo(final String test) {
> -//                     return "test".contentEquals(test)
> -//                             || "test2".contentEquals(test);
> +//                     return test.contentEquals("test")
> +//                             || test.contentEquals("test2");
>  //                  }
>                }
> \ No newline at end of file
> ```
> 
> ```
>     expected: 
>       "              import java.util.stream.Stream;
>                     public class A {
>                         {
>                             String s = null;
>                             if("test".equals(s)) {}
>                             if("test".equalsIgnoreCase(s)) {}
>                             System.out.println("test".compareTo(s));
>                             System.out.println("test".compareToIgnoreCase(s));
>                             System.out.println("test".contentEquals(s));
>                             System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
>                         }
>       //                  boolean isFoo(final String test) {
>       //                     return "test".contentEquals(test)
>       //                             || "test2".contentEquals(test);
>       //                  }
>                     }"
>      but was: 
>       "              import java.util.stream.Stream;
>                     public class A {
>                         {
>                             String s = null;
>                             if("test".equals(s)) {}
>                             if("test".equalsIgnoreCase(s)) {}
>                             System.out.println("test".compareTo(s));
>                             System.out.println("test".compareToIgnoreCase(s));
>                             System.out.println("test".contentEquals(s));
>                             System.out.println(Stream.of("test").filter(item -> "test".contentEquals(item)).findFirst());
>                         }
>       //                  boolean isFoo(final String test) {
>       //                     return test.contentEquals("test")
>       //                             || test.contentEquals("test2");
>       //                  }
>                     }"
>         at app//org.openrewrite.test.RewriteTest.assertContentEquals(RewriteTest.java:620)
>         at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:509)
>         at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:132)
>         at app//org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:127)
>         at app//org.openrewrite.staticanalysis.EqualsAvoidsNullTest.invertConditional(EqualsAvoidsNullTest.java:36)
> ```

Right thanks, now I have provided an positive failing test.

File: src/test/java/org/openrewrite/staticanalysis/EqualsAvoidsNullTest.java [hunk 2]
```diff
@@ -37,26 +37,68 @@ void invertConditional() {
...
+                  }
+              }
+
+              public static class B {
+                  String getBar(final String test) {
+                      return null;
+                  }
+
+                  static String getBar2(final String test) {
+                      return null;
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/EqualsAvoidsNullTest.java:68]: Instead of expanding the small document example it's probably best to start a new unit test for the exact problem you want to replicate. We try to keep the document example small for that precise purpose: show a new issue in isolation, and only that particular case, as that then also helps stepping through with the debugger.

**Comment by timtebeek** [comment]
Hi @punkratz312 Could you provided some context to go along with your review request? You're changing a unit test, and now that unit test is not making any changes anymore, presumably because the recipe only looks at literals, not constants. Correct?

The associated failure is
```
EqualsAvoidsNullTest > invertConditional() FAILED
    java.lang.AssertionError: Recipe was expected to make a change but made no changes.
        at org.openrewrite.test.LargeSourceSetCheckingExpectedCycles.afterCycle(LargeSourceSetCheckingExpectedCycles.java:119)
...
        at org.openrewrite.staticanalysis.EqualsAvoidsNullTest.invertConditional(EqualsAvoidsNullTest.java:36)
```

Do you plan to expand this recipe on this PR to cover that case as well? Or do you need input of some form first?

**Comment by punkratz312** [comment]
Indeed, the field allocation setup isn’t sufficient. I don’t fully understand it yet, but I need to deliver a feature to production soon, so I don’t have much time right now. Could you give me a hint? It’s just a snippet we need to adjust.

We need to extend the test to include the default case, covering both states. While I agree with you on keeping a simple, clear example (KISS principle) at the top, it’s also important to maintain high test coverage and variation in our regression tests to capture edge cases.

Thanks!

**Comment by punkratz312** [comment]
Even after removing all other checks except the method matcher itself, the filtering works quite well. However, the code responsible for making the adjustments isn't consistently applying the swap to constants.

**Comment by timtebeek** [comment]
So right now the tests are not matching since the first argument is no longer a String literal:
https://github.com/openrewrite/rewrite-static-analysis/blob/8f3a735012f0c89cd09e4b2a00150f35a95de9a5/src/main/java/org/openrewrite/staticanalysis/EqualsAvoidsNullVisitor.java#L62

Instead, the first argument is a field reference; if we want the recipe to make changes for those too (and only if constant, not nullable) then we should update the logic to account for those cases too, in addition to the String literals evaluated previously. I'm not sure that's as easily done here when those referenced fields are external, although we've [done similar before](https://github.com/openrewrite/rewrite/blob/a98d83d698dda9971c93a0767b7f85d5866f306d/rewrite-java/src/main/java/org/openrewrite/java/ReplaceStringLiteralWithConstant.java#L153-L157). Either way it would be good to have separate tests: one for literals, and one for fields.

**Comment by punkratz312** [comment]
@timtebeek see you tomorrow 👍 09.00

**Comment by punkratz312** [comment]
Thanks for mentioning me! I had to deliver a feature, but I watched the script. Thanks for the demo.

**Comment by timtebeek** [comment]
> Thanks for mentioning me! I had to deliver a feature, but I watched the script. Thanks for the demo.

Of course! Thanks for helping out; look forward to working together even more. :)

**Comment by punkratz312** [comment]
The expression seems correct to me, so I'm unsure why the change isn't being applied. Additionally, I've noticed that `KeepOrderForStringVsString` is also an issue, but it’s currently disabled as it’s not critical.

In production, it's also not detecting these cases. Since the expression seems right, perhaps the issue lies beyond this class—I'm not sure. Feel free to take over, please. @timtebeek

Cheers and best wishes to my friends near Amster- and Rotterdam!  
I miss the nice, fresh sea air so much, especially while riding around town on my bike—ring ring! 🔔 

image

**Comment by knutwannheden** [comment]
> Could you please take a look at this? It might require minimal effort to fix my confusion or the bug, as it seems to be a low-hanging fruit.

I couldn't find any failing tests. I merged in the latest main as the CI build job failed for some reason.

**Comment by knutwannheden** [comment]
> I couldn't find any failing tests. I merged in the latest main as the CI build job failed for some reason.

My bad. The checkout failed because I had some local changes...

**Comment by punkratz312** [comment]
the test `constantList` is the one not working in production, imho.

**Comment by punkratz312** [comment]
https://github.com/openrewrite/rewrite-static-analysis/pull/390/files
